src/cozmo/nav_memory_map.py

```diff
@@ -0,0 +1,316 @@
+    ClearOfCliff = _entry_type("ClearOfCliff", _clad_enum.ClearOfCliff)
+
+    #: The node contains a :class:`~cozmo.objects.LightCube`.
+    ObstacleCube = _entry_type("ObstacleCube", _clad_enum.ObstacleCube)
+
+    #: The node contains a :class:`~cozmo.objects.Charger`.
+    ObstacleCharger = _entry_type("ObstacleCharger", _clad_enum.ObstacleCharger)
+
+    #: The node contains a cliff (a sharp drop).
+    Cliff = _entry_type("Cliff", _clad_enum.Cliff)
+
+    #: The node contains a visible edge (based on the camera feed).
+    VisionBorder = _entry_type("VisionBorder", _clad_enum.VisionBorder)
+
+
+NodeContentTypes._init_class()
+
+
+class NavMemoryMapGridNode:
+    """A node in a :class:`NavMemoryMapGrid`.
+
+    Leaf nodes contain content, all other nodes are split into 4 equally sized
+    children.
+
+    Child node indices are stored in the following X,Y orientation:
+
+        +---+----+---+
+        | ^ | 2  | 0 |
+        +---+----+---+
+        | Y | 3  | 1 |
+        +---+----+---+
+        |   | X->|   |
+        +---+----+---+
+    """
+    def __init__(self, depth, size, center, parent):
+        #: int: The depth of this node. I.e. how far down the quad-tree is it.
+        self.depth = depth
+
+        #: float: The size (width or length) of this square node.
+        self.size = size
+
+        #: :class:`~cozmo.util.Vector3`: The center of this node.
+        self.center = center  # type: util.Vector3
+
+        #: :class:`NavMemoryMapGridNode`: The parent of this node. Is ``None`` for the root node.
+        self.parent = parent  # type: NavMemoryMapGridNode
+
+        #: list of :class:`NavMemoryMapGridNode`: ``None`` for leaf nodes, a list of 4
+        #: child nodes otherwise.
+        self.children = None
+
+        #: An attribute of :class:`NodeContentTypes`: The content type in this
+        #: node. Only leaf nodes have content, this is ``None`` for all other
+        #: nodes.
+        self.content = None  # type: _NodeContentType
+
+        self._next_child = 0  # Used when building to track which branch to follow (
+
+    def __repr__(self):
+        return '<%s center: %s size: %s content: %s>' % (
+            self.__class__.__name__, self.center, self.size, self.content)
+
+    def contains_point(self, x, y):
+        """Test if the node contains the given x,y coordinates.
+
+        Args:
+            x (float): x coordinate for the point
+            y (float): y coordinate for the point
+
+        Returns:
+            bool: True if the node contains the point, False otherwise.
+        """
+        half_size = self.size * 0.5
+        dist_x = abs(self.center.x - x)
+        dist_y = abs(self.center.y - y)
+        return (dist_x <= half_size) and (dist_y <= half_size)
+
+    def get_node(self, x, y, _assumed_in_bounds=False):
+        """Get the node at the given x,y coordinates.
+
+        Args:
+            x (float): x coordinate for the point
+            y (float): y coordinate for the point
+
```

Comment:
I didn't want to expose that as it's an implementation detail, but I'll rework it to use a private method to hide it a bit better.